nixos/modules/services/cluster/kubernetes/controller-manager.nix

```diff
@@ -1,109 +1,64 @@
-    };
-
-    clusterCidr = mkOption {
-      description = lib.mdDoc "Kubernetes CIDR Range for Pods in cluster.";
-      default = top.clusterCidr;
-      defaultText = literalExpression "config.${otop.clusterCidr}";
-      type = str;
-    };
-
-    enable = mkEnableOption (lib.mdDoc "Kubernetes controller manager");
-
-    extraOpts = mkOption {
-      description = lib.mdDoc "Kubernetes controller manager extra command line options.";
-      default = "";
-      type = separatedString " ";
-    };
-
-    featureGates = mkOption {
-      description = lib.mdDoc "List set of feature gates";
-      default = top.featureGates;
-      defaultText = literalExpression "config.${otop.featureGates}";
-      type = listOf str;
-    };
-
-    kubeconfig = top.lib.mkKubeConfigOptions "Kubernetes controller manager";
-
-    leaderElect = mkOption {
-      description = lib.mdDoc "Whether to start leader election before executing main loop.";
-      type = bool;
-      default = true;
-    };
-
-    rootCaFile = mkOption {
-      description = lib.mdDoc ''
-        Kubernetes controller manager certificate authority file included in
-        service account's token secret.
-      '';
-      default = top.caFile;
-      defaultText = literalExpression "config.${otop.caFile}";
-      type = nullOr path;
-    };
-
-    securePort = mkOption {
-      description = lib.mdDoc "Kubernetes controller manager secure listening port.";
-      default = 10252;
-      type = int;
-    };
-
-    serviceAccountKeyFile = mkOption {
-      description = lib.mdDoc ''
-        Kubernetes controller manager PEM-encoded private RSA key file used to
-        sign service account tokens
-      '';
-      default = null;
-      type = nullOr path;
-    };
-
-    tlsCertFile = mkOption {
-      description = lib.mdDoc "Kubernetes controller-manager certificate file.";
-      default = null;
-      type = nullOr path;
-    };
-
-    tlsKeyFile = mkOption {
-      description = lib.mdDoc "Kubernetes controller-manager private key file.";
-      default = null;
-      type = nullOr path;
-    };
-
-    verbosity = mkOption {
-      description = lib.mdDoc ''
-        Optional glog verbosity level for logging statements. See
-        <https://github.com/kubernetes/community/blob/master/contributors/devel/logging.md>
+  options.services.kubernetes.controllerManager = with types; {
+    enable = mkEnableOption (mdDoc "Kubernetes controller manager");
+
+    settings = mkOption {
+      description = ''
+        Configuration for kube-controller-manager, see:
+          <https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager>.
+        All attrs defined here translates directly to flags of syntax `--<name>="<value>"`
+        which is provided as command line argument to the kube-controller-manager binary.
       '';
-      default = null;
-      type = nullOr int;
+      type = types.submodule {
+        freeformType = attrsOf (oneOf [
```

Comment:
Before that we had a typed API for a lot of stuff with validation on eval time for free for that. Wouldn't it make sense to add what existed before in here and work with `mkRenamedOptionModule`? This has two advantages: first, we keep the nice properties of validation from the module system and second it'd be way easier for folks to migrate their configuration (right now this will be rather painful).

Reply:
While I agree that the eval time checks are nice, it isn't exactly for free - the cost is maintenance in nixpkgs, which should be balanced. And the sad truth is that all the many options in here didn't get maintained. For this reason, there were options here which had been completely removed by upstream (we didn't see it because our tests weren't touching them), and option defaults which differed from upstream defaults. I know I'm co-maintainer of this package/module, so that's partly my fault.

Unicorns and Rainbows: What I would like to work for with Kubernetes upstream instead, is to get `--config-check` flags on all binaries, which we can use for config validation rather than implementing it ourselves in nixpkgs.

Until we get unicorns: Can we perhaps add _some_ of the pre-existing options without re-adding _all_ of them? (to make the maintenance burden easier). E.g. Particularly those with enum types and stricter type checks than just strings are most valuable, I guess?

Reply:
I think it's OK to find a middle ground here for sure, I agree with both sentiments. In particularly, it's not so much the complex options I'd like to see have some settings and defaults, but the "simple hardening ones", eg. I think we should re-harden bind addresses to avoid participating in the sad statistics of open-to-the-world apiservers etc :)

I don't think a lot of the others are _that_ difficult to port for people, following the release notes.